In [1]:
using Flux, MLDataPattern, Mill, JsonGrinder, JSON, HierarchicalUtils, StatsBase
using JsonGrinder: DictEntry, Entry

load files in data/documents and parse them

In [2]:
data_dir = "../../../data/documents"
sch = JsonGrinder.schema(readdir(data_dir, join=true), x->open(JSON.parse, x))

[Dict] 	# updated = 16
  ├───── metadata: [Dict] 	# updated = 16
  │                  ├── authors: [List] 	# updated = 16
  │                  │              ⋮
  │                  └──── title: [Scalar - String], 14 unique values 	# updated = 16
  ├── ref_entries: [Dict] 	# updated = 16
  │                  ├── TABREF4: [Dict] 	# updated = 3
  │                  │              ⋮
  │                  ├── FIGREF3: [Dict] 	# updated = 3
  │                  │              ⋮
  │                  ⋮
  │                  └── TABREF2: [Dict] 	# updated = 6
  │                                 ⋮
  ⋮
  └───── abstract: [List] 	# updated = 16
                     └── [Dict] 	# updated = 22
                           ⋮

suggest default extractor with some keys as field

In [3]:
extractor = suggestextractor(sch)

┌ Warning: [:bib_entries][:BIBREF33]: key other_ids contains empty array, skipping
└ @ JsonGrinder ~/work/JsonGrinder.jl/JsonGrinder.jl/src/schema/dict.jl:62
┌ Warning: [:bib_entries][:BIBREF127]: key authors contains empty array, skipping
└ @ JsonGrinder ~/work/JsonGrinder.jl/JsonGrinder.jl/src/schema/dict.jl:62
┌ Warning: [:bib_entries][:BIBREF127]: key other_ids contains empty array, skipping
└ @ JsonGrinder ~/work/JsonGrinder.jl/JsonGrinder.jl/src/schema/dict.jl:62
┌ Warning: [:bib_entries][:BIBREF67]: key other_ids contains empty array, skipping
└ @ JsonGrinder ~/work/JsonGrinder.jl/JsonGrinder.jl/src/schema/dict.jl:62
┌ Warning: [:bib_entries][:BIBREF79]: key other_ids contains empty array, skipping
└ @ JsonGrinder ~/work/JsonGrinder.jl/JsonGrinder.jl/src/schema/dict.jl:62
┌ Warning: [:bib_entries][:BIBREF153]: key authors contains empty array, skipping
└ @ JsonGrinder ~/work/JsonGrinder.jl/JsonGrinder.jl/src/schema/dict.jl:62
┌ Warning: [:bib_entries][:BIBREF153]: key other_ids 

Dict
  ├───── metadata: Dict
  │                  ├── authors: Array of
  │                  │              ⋮
  │                  └──── title: Categorical d = 15
  ├── ref_entries: Dict
  │                  ├── TABREF5: Dict
  │                  │              ⋮
  │                  ├── TABREF4: Dict
  │                  │              ⋮
  │                  ⋮
  │                  └── TABREF2: Dict
  │                                 ⋮
  ⋮
  └───── abstract: Array of
                     └── Dict
                           ⋮

show whole extractor

In [4]:
printtree(extractor)

Dict
  ├───── metadata: Dict
  │                  ├── authors: Array of
  │                  │              └── Dict
  │                  │                    ├─────── middle: Array of
  │                  │                    │                  └── Categorical d = 13
  │                  │                    ├──────── first: Categorical d = 87
  │                  │                    ├─────── suffix: Categorical d = 2
  │                  │                    ├───────── last: Categorical d = 80
  │                  │                    ├──────── email: Categorical d = 5
  │                  │                    └── affiliation: Dict
  │                  │                                       ├─── laboratory: Categorical d = 6
  │                  │                                       ├───── location: Dict
  │                  │                                       │                  ├── settlement: Categorical d = 10
  │                  │                                       │ 

we see that there are some dictionaries with lots of keys, let's examine schema
list_lens lets us iterate over all elements in a way we know their position in schema
this prints lengths of children of all dict entries.

In [5]:
for i in list_lens(sch)
    e = get(sch, i)
    if e isa DictEntry
        @info i length(e.childs)
    end
end

┌ Info: (@lens _)
└   length(e.childs) = 7
┌ Info: (@lens _.childs[:metadata])
└   length(e.childs) = 2
┌ Info: (@lens _.childs[:metadata].childs[:authors].items)
└   length(e.childs) = 6
┌ Info: (@lens _.childs[:metadata].childs[:authors].items.childs[:affiliation])
└   length(e.childs) = 3
┌ Info: (@lens _.childs[:metadata].childs[:authors].items.childs[:affiliation].childs[:location])
└   length(e.childs) = 5
┌ Info: (@lens _.childs[:ref_entries])
└   length(e.childs) = 13
┌ Info: (@lens _.childs[:ref_entries].childs[:TABREF4])
└   length(e.childs) = 3
┌ Info: (@lens _.childs[:ref_entries].childs[:FIGREF3])
└   length(e.childs) = 2
┌ Info: (@lens _.childs[:ref_entries].childs[:FIGREF2])
└   length(e.childs) = 2
┌ Info: (@lens _.childs[:ref_entries].childs[:FIGREF4])
└   length(e.childs) = 2
┌ Info: (@lens _.childs[:ref_entries].childs[:TABREF3])
└   length(e.childs) = 3
┌ Info: (@lens _.childs[:ref_entries].childs[:TABREF5])
└   length(e.childs) = 3
┌ Info: (@lens _.childs[:ref_entr

that's a lots of numbers, let's see histogram

In [6]:
length_hist = StatsBase.countmap([length(get(sch, i).childs) for i in list_lens(sch) if get(sch, i) isa DictEntry])

Dict{Int64, Int64} with 12 entries:
  5   => 1
  8   => 12
  1   => 43
  0   => 59
  6   => 1
  9   => 91
  3   => 9
  7   => 1
  103 => 1
  4   => 97
  13  => 1
  2   => 8

we see highest lengths are 103 and 13, let's set 13 as a threshold

In [7]:
extractor = suggestextractor(sch, (; key_as_field=13))

[ Info: [:ref_entries] seems to store values in keys, therefore node is treated as bag with keys as extra values.
[ Info: [:bib_entries] seems to store values in keys, therefore node is treated as bag with keys as extra values.
┌ Warning: [:abstract]: key cite_spans contains empty array, skipping
└ @ JsonGrinder ~/work/JsonGrinder.jl/JsonGrinder.jl/src/schema/dict.jl:62
┌ Warning: [:abstract]: key ref_spans contains empty array, skipping
└ @ JsonGrinder ~/work/JsonGrinder.jl/JsonGrinder.jl/src/schema/dict.jl:62


Dict
  ├───── metadata: Dict
  │                  ├── authors: Array of
  │                  │              ⋮
  │                  └──── title: Categorical d = 15
  ├── ref_entries: KeyAsField
  │                  ├── String
  │                  └── Dict
  │                        ⋮
  ⋮
  └───── abstract: Array of
                     └── Dict
                           ⋮

show new extractor

In [8]:
printtree(extractor)

Dict
  ├───── metadata: Dict
  │                  ├── authors: Array of
  │                  │              └── Dict
  │                  │                    ├─────── middle: Array of
  │                  │                    │                  └── Categorical d = 13
  │                  │                    ├──────── first: Categorical d = 87
  │                  │                    ├─────── suffix: Categorical d = 2
  │                  │                    ├───────── last: Categorical d = 80
  │                  │                    ├──────── email: Categorical d = 5
  │                  │                    └── affiliation: Dict
  │                  │                                       ├─── laboratory: Categorical d = 6
  │                  │                                       ├───── location: Dict
  │                  │                                       │                  ├── settlement: Categorical d = 10
  │                  │                                       │ 

this extractor looks much better
but still, some values are very sparse,
let's print all parts of schema where each value is observed only once

In [9]:
for i in list_lens(sch)
    e = get(sch, i)
    if e isa Entry && maximum(values(e.counts)) == 1
        @info i
    end
end

[ Info: (@lens _.childs[:metadata].childs[:authors].items.childs[:middle].items)
[ Info: (@lens _.childs[:metadata].childs[:authors].items.childs[:affiliation].childs[:location].childs[:region])
[ Info: (@lens _.childs[:ref_entries].childs[:TABREF4].childs[:text])
[ Info: (@lens _.childs[:ref_entries].childs[:TABREF4].childs[:html])
[ Info: (@lens _.childs[:ref_entries].childs[:FIGREF3].childs[:text])
[ Info: (@lens _.childs[:ref_entries].childs[:FIGREF2].childs[:text])
[ Info: (@lens _.childs[:ref_entries].childs[:FIGREF4].childs[:text])
[ Info: (@lens _.childs[:ref_entries].childs[:TABREF3].childs[:text])
[ Info: (@lens _.childs[:ref_entries].childs[:TABREF3].childs[:html])
[ Info: (@lens _.childs[:ref_entries].childs[:TABREF5].childs[:text])
[ Info: (@lens _.childs[:ref_entries].childs[:TABREF5].childs[:html])
[ Info: (@lens _.childs[:ref_entries].childs[:TABREF6].childs[:type])
[ Info: (@lens _.childs[:ref_entries].childs[:TABREF6].childs[:text])
[ Info: (@lens _.childs[:ref_entrie

 we can see lots of leaves under `bib_entries`, which is cased by uniqueness of keys here
but apart from that, we can see other interesting fields
[ Info: (@lens _.childs[:metadata].childs[:authors].items.childs[:middle].items)
[ Info: (@lens _.childs[:metadata].childs[:authors].items.childs[:last])
[ Info: (@lens _.childs[:metadata].childs[:authors].items.childs[:affiliation].childs[:location].childs[:region])
[ Info: (@lens _.childs[:paper_id])
[ Info: (@lens _.childs[:body_text].items.childs[:text])
[ Info: (@lens _.childs[:body_text].items.childs[:ref_spans].items.childs[:start])
[ Info: (@lens _.childs[:body_text].items.childs[:ref_spans].items.childs[:end])
[ Info: (@lens _.childs[:back_matter].items.childs[:text])
[ Info: (@lens _.childs[:back_matter].items.childs[:cite_spans].items.childs[:ref_id])
[ Info: (@lens _.childs[:back_matter].items.childs[:cite_spans].items.childs[:start])
[ Info: (@lens _.childs[:back_matter].items.childs[:cite_spans].items.childs[:text])
[ Info: (@lens _.childs[:back_matter].items.childs[:cite_spans].items.childs[:end])
[ Info: (@lens _.childs[:back_matter].items.childs[:ref_spans].items.childs[:start])
[ Info: (@lens _.childs[:back_matter].items.childs[:ref_spans].items.childs[:text])
[ Info: (@lens _.childs[:back_matter].items.childs[:ref_spans].items.childs[:end])

let's remove some of them from extractor

In [10]:
delete!(extractor.dict, :paper_id)
delete!(extractor.dict[:metadata].dict[:authors].item.dict, :last)
delete!(extractor.dict[:metadata].dict[:authors].item.dict, :middle)

Dict{Symbol, JsonGrinder.AbstractExtractor} with 4 entries:
  :first       => ExtractCategorical
  :suffix      => ExtractCategorical
  :email       => ExtractCategorical
  :affiliation => ExtractDict

we can also notice, that some long texts are extracted as categorical variables, e.g.

In [11]:
extractor[:body_text].item[:text]
extractor[:body_text].item[:section]

String

let's replace them manually by string extractors
note that we need to use the .dict, as the [] accessor on item is just readonly syntax-sugar

In [12]:
extractor[:body_text].item.dict[:text] = ExtractString()
extractor[:body_text].item.dict[:section] = ExtractString()

String

this concludes example about examining schema and modifying extractor accordingly.

In [13]:
using JsonGrinder: is_intable, is_floatable, unify_types, extractscalar
function string_multi_representation_scalar_extractor()
	vcat([
	(e -> unify_types(sch[:paper_id]) <: String,
		(e, uniontypes) -> MultipleRepresentation((
			ExtractCategorical(top_n_keys(e, 20), uniontypes),
			extractscalar(unify_types(e), e, uniontypes)
		)))
	], JsonGrinder.default_scalar_extractor())
end

top_n_keys(e::Entry, n::Int) = map(x->x[1], sort(e.counts |> collect, by=x->x[2], rev=true)[begin:min(n, end)])
suggestextractor(sch, (;
	scalar_extractors=string_multi_representation_scalar_extractor(),
	key_as_field=13,
	)
) |> printtree
unify_types(sch[:paper_id]) <: String

[ Info: [:ref_entries] seems to store values in keys, therefore node is treated as bag with keys as extra values.
[ Info: [:bib_entries] seems to store values in keys, therefore node is treated as bag with keys as extra values.
┌ Warning: [:abstract]: key cite_spans contains empty array, skipping
└ @ JsonGrinder ~/work/JsonGrinder.jl/JsonGrinder.jl/src/schema/dict.jl:62
┌ Warning: [:abstract]: key ref_spans contains empty array, skipping
└ @ JsonGrinder ~/work/JsonGrinder.jl/JsonGrinder.jl/src/schema/dict.jl:62
Dict
  ├───── metadata: Dict
  │                  ├── authors: Array of
  │                  │              └── Dict
  │                  │                    ├─────── middle: Array of
  │                  │                    │                  └── MultiRepresentation
  │                  │                    │                        ├── e1: Categorical d = 13
  │                  │                    │                        └── e2: String
  │                  │               

true

---

*This notebook was generated using [Literate.jl](https://github.com/fredrikekre/Literate.jl).*